server/api/cvrs.py

```diff
@@ -782,20 +783,93 @@ def join_ballots_to_interpretations(
 def parse_hart_cvrs(
     jurisdiction: Jurisdiction,
 ) -> Tuple[CVR_CONTESTS_METADATA, Iterable[CvrBallot]]:
-    # Hart CVRs consist of a zip file containing an individual XML file for each
-    # ballot's CVR. Our parsing steps:
-    # 1. Unzip the files
-    # 2. First, parse out the contest and choice names. We have to do this in
-    #    a separate pass since our storage scheme for interpretations requires
-    #    knowing all of the contest and choice names up front.
-    # 3. Next, parse out the interpretations.
+    # Hart CVRs consist of a ZIP file containing an individual XML file for each ballot's CVR.
+    # Separate from this ZIP file, an optional scanned ballot information CSV can be provided. If
+    # the latter is provided, the `UniqueIdentifier`s in it will be used as imprinted IDs.
+    # Otherwise, `CvrGuid`s will be used.
+    #
+    # Our parsing steps:
+    # 1. Unzip the wrapper ZIP file.
+    # 2. Expect either [ another ZIP file ] or [ another ZIP file and a CSV ].
+    # 3. If a CSV is found, parse it as a scanned ballot information CSV.
+    # 4. Unzip the CVRs ZIP file.
+    # 5. Parse out the contest and choice names. We have to do this in a separate pass since our
+    #    storage scheme for interpretations requires knowing all of the contest and choice names up
+    #    front.
+    # 3. Then, parse out the interpretations.
+
+    wrapper_zip_file = retrieve_file(jurisdiction.cvr_file.storage_path)
+    files = unzip_files(wrapper_zip_file)
+    wrapper_zip_file.close()
+
+    cvrs_zip_file: Union[BinaryIO, None] = None
+    scanned_ballot_information_file: Union[BinaryIO, None] = None
+    for name, file in files.items():
+        if name.lower().endswith(".zip"):
+            cvrs_zip_file = file
+        if name.lower().endswith(".csv"):
+            scanned_ballot_information_file = file
+
+    if cvrs_zip_file is None:
+        raise UserError("Expected a file with a .zip extension.")
+
+    def construct_cvr_guid_to_unique_identifier_mapping(
+        scanned_ballot_information_file: BinaryIO,
+    ) -> Dict[str, str]:
+        validate_not_empty(scanned_ballot_information_file)
+        text_file = decode_csv(scanned_ballot_information_file)
+
+        # Determine whether the scanned ballot information CSV has a format version row and skip it
+        # if so
+        has_format_version_row = False
+        first_line = text_file.readline()
+        if first_line.startswith("#FormatVersion") or "," not in first_line:
```

Comment:
It's very possible that every scanned ballot information CSV has this format version row, but I didn't want to make that assumption

Reply:
I've generally been taking the approach that since all of these files are auto-generated, they will look exactly how the sample files look unless I see an actual counterexample. That way I don't complicate the code trying to account for the endless number of variations that might exist and just focus on the ones we actually encounter. So I think it's fine to assume the format version row exists. That being said, not a big deal either way

Reply:
Really good to know as a principle that these files are generated and not hand modified! Updated the code to assume that the format version row exists: https://github.com/votingworks/arlo/pull/1558/commits/72d2bbf0cee680adf9d426afa2bc169f4cc26caf